Implement CSE for dynamo guards.

ysiraichi opened the pull request:
This PR extracted the CSE part of the code in #89707.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

ysiraichi committed:
First CSE implementation.

ysiraichi commented:
In summary, it implements the CSE in 3 steps:

1. Parse each guard expression into an AST
2. Traverse each parsed AST, counting the number of occurrences at each level
    - Holds a map from the stringified node (using `ast.unparse` function) to a counter
3. Traverse each parsed AST, replacing the nodes that occur more than a threshold (currently, 2) by a new variable

**Returns:**
    - the assignments of each new variable to its corresponding expression 
    - the new guard expressions

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -750,7 +920,27 @@ def compile_check_fn(
+            preface = [
+                "try:",
+                *[indent(line) for line in preface],
+                "except:",
```
Do you know why you need the try-except here? It doesn't seem like it should be necessary

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -611,6 +642,145 @@ def _produce_guard_code(
+            if _ast_unparse_implemented(node) and not isinstance(
+                node, PyExprCSEPass.IGNORED_NODE_TYPES
+            ):
+                expr = _ast_unparse(node)
```
NB: because visit recursively traverses children, this is quadratic in expression size

ezyang commented:
Perf run kicked off on https://github.com/pytorch/pytorch/actions/runs/4638385442

ezyang commented:
This is a very nice improvement, worth 2-4%

<img width="769" alt="image" src="https://user-images.githubusercontent.com/13564/230692267-5eb4355b-7f7f-4bdb-9273-3396fff0b769.png">

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -611,6 +642,145 @@ def _produce_guard_code(
+                        node_ = super().visit(node)
+                        expr_ = _ast_unparse(node_)
+                        var_name = self._gen_name()
+                        self._preface.append(f"{var_name} = {expr_}")
```
Hoisting the expressions all the way to the beginning is not sound in general, and I think this explains the need for the try...catch.

The order we run guards matters. Suppose that we have a guard on x being tensor, and another guard on x.grad. We must test that x is tensor, before we attempt to access x.grad. With the hoisting here, you may violate the dependencies on guard accesses.

ysiraichi replied:
> I think this explains the need for the try...catch.
Yes.

ezyang reviewed (changes requested):
Need to preserve expression ordering

voznesenskym commented on torch/_dynamo/guards.py:
```diff
@@ -611,6 +642,145 @@ def _produce_guard_code(
+
+    def _new_var(self, prefix: str = "_var") -> str:
+        name = f"{prefix}{self._counter}"
+        self._counter += 1
```
When /why do we need to increment? do all new vars just get a number assigned?

ezyang replied:
This is assigning names to each CSE'd expr, so obvs they need different names

voznesenskym replied:
> So, I think I can make this PR better (I think I can finish the implementation tomorrow) by:
> 
> 1. We can make it work without `ast.unparse` (eliminating the quadratic time) by using the column information from parsing
> 2. Use `ALLOW_NODE_TYPES` (the inverse of what we have now) so as to filter the nodes interesting to the CSE, namely: `Subscript`, `Attribute`, and `Call` (for `getattr` calls)
> 3. Divide the guard function into non-ShapeEnv guards and ShapeEnv guards (which will free us from the try-catch)

We already divide ordering, shape_env guards get added separately. It would not be a stretch to do this.

voznesenskym replied:
Right, but I was wondering if we can use the prior unique names, there should be no collisions.

ysiraichi replied:
The first version of CSE (in the other PR) did exactly that. The problems were:

- We still needed the counter, since the expr-var mapping wasn't injective (i.e. collisions could happen)
- Increased compilation time
- Ad-hoc logic for replacing prohibited symbols

voznesenskym replied:
> (FYI: I will be off until May 1)

If need be, I can help finish this PR and land it under your attribution.

ysiraichi commented on torch/_dynamo/guards.py:
```diff
@@ -611,6 +642,145 @@ def _produce_guard_code(
+#        'v' will be defined in the 'preface' list (output argument to
+#        'NodeTransformer')
+class PyExprCSEPass:
+    IGNORED_NODE_TYPES = (
```
It was done in an ad-hoc way. Basically, I went through the list of node types and decided whether they were interesting for CSE purposes. Now that I think about it, maybe I should have created the a list of `ALLOWED_NODE_TYPES`.

ysiraichi commented:
So, I think I can make this PR better (I think I can finish the implementation tomorrow) by:

1. We can make it work without `ast.unparse` (eliminating the quadratic time) by using the column information from parsing
2. Use `ALLOW_NODE_TYPES` (the inverse of what we have now) so as to filter the nodes interesting to the CSE, namely: `Subscript`, `Attribute`, and `Call` (for `getattr` calls)
3. Divide the guard function into non-ShapeEnv guards and ShapeEnv guards (which will free us from the try-catch)

@ezyang 
What do you think?

voznesenskym commented:
> 3. Divide the guard function into non-ShapeEnv guards and ShapeEnv guards (which will free us from the try-catch)

We have prior art for this already, in terms of treating shape_env code separately from other added guard code.

ezyang commented:
> We can make it work without ast.unparse (eliminating the quadratic time) by using the column information from parsing

By using the original string? That's seems brittle. According to the perf run your change doesn't materially change compile time, so I think the quadratic term doesn't matter, leave it as is (with a comment saying that it's quadratic.)

> Use ALLOW_NODE_TYPES (the inverse of what we have now) so as to filter the nodes interesting to the CSE, namely: Subscript, Attribute, and Call (for getattr calls)

Sure, though IMO it's not a big deal.

> Divide the guard function into non-ShapeEnv guards and ShapeEnv guards (which will free us from the try-catch)

As Voz says, this is the minimum you will need to do, but I don't think it is enough. Imagine something like:

```
isinstance(x, list)
len(x) == 2
isinstance(x[0], list)
len(x[0]) == 2
isinstance(x[0][0], int)
isinstance(x[0][1], int)
isinstance(x[1], int)
```

Naive CSE will move x[0] access before we have confirmed that x is a list.

ezyang commented:
I think a useful thing to know is what the performance difference between:

```
return a and b and c
```

vs

```
if not a:
  return False
if not b:
  return False
if not c:
  return False
return True
```

is. If there is not much difference, then the easiest thing to do is to stop compiling guards as a big list of `and` statements and then you can CSE at the first point a subexpression is used, without requiring to reorder things.

ysiraichi commented:
> According to the perf run your change doesn't materially change compile time, so I think the quadratic term doesn't matter

Not sure what you mean by that. By compile time, do you mean after we introduce JIT to the guards?

ysiraichi commented:
Find below the benchmark (torchbench) results for the 2 versions @ezyang mentioned. In summary:

- They don't seem to be significantly different
- There was 1 benchmark that got >1.1x speedup: `resnext50_32x4d`
- There were 2 benchmarks that got <0.9x speedup: `dlrm` (0.88x) and `hf_T5_large` (0.81x)
- There were 3 benchmarks that got <0.98x speedup: `hf_T5`, `hf_T5_base`, and `mobilenet_v2`

<details>
<summary> Raw Results </summary>

- `baseline` refers to the commit `1d7133c542f` (February 15)
- `if-chain` refers to the second version (literally, a chain of ifs)

| Name                           | Latency (baseline) | Latency (if-chain) | Speedup |
|--------------------------------|--------------------|--------------------|---------|
| BERT_pytorch                   | 38.2049            | 38.1404            | 1.0017  |
| Background_Matting             | 148.4077           | 148.1106           | 1.0020  |
| LearningToPaint                | 42.1799            | 42.1211            | 1.0014  |
| Super_SloMo                    | 319.2586           | 319.4182           | 0.9995  |
| alexnet                        | 34.7081            | 34.6411            | 1.0019  |
| dcgan                          | 25.2886            | 25.2632            | 1.0010  |
| densenet121                    | 190.7768           | 192.6175           | 0.9904  |
| dlrm                           | 2.0278             | 2.2835             | 0.8880  |
| drq                            | 0.7249             | 0.7233             | 1.0022  |
| fastNLP_Bert                   | 17.2611            | 17.1888            | 1.0042  |
| functorch_dp_cifar10           | 20.6523            | 20.6971            | 0.9978  |
| functorch_maml_omniglot        | 0.6388             | 0.6464             | 0.9882  |
| hf_Albert                      | 15.0872            | 15.0448            | 1.0028  |
| hf_Bart                        | 18.9939            | 18.8683            | 1.0067  |
| hf_Bert                        | 17.6168            | 17.9100            | 0.9836  |
| hf_DistilBert                  | 8.0332             | 8.0316             | 1.0002  |
| hf_GPT2                        | 37.0473            | 37.0166            | 1.0008  |
| hf_GPT2_large                  | 199.5633           | 199.7220           | 0.9992  |
| hf_Reformer                    | 15.5395            | 15.6129            | 0.9953  |
| hf_T5                          | 90.9850            | 98.2737            | 0.9258  |
| hf_T5_base                     | 305.4032           | 319.5407           | 0.9558  |
| hf_T5_large                    | 127.2385           | 156.5839           | 0.8126  |
| lennard_jones                  | 0.2202             | 0.2198             | 1.0018  |
| maml                           | 114.7211           | 116.2960           | 0.9865  |
| maml_omniglot                  | 0.8444             | 0.8205             | 1.0291  |
| mnasnet1_0                     | 19.5482            | 19.5491            | 1.0000  |
| mobilenet_v2                   | 12.3285            | 12.9194            | 0.9543  |
| mobilenet_v3_large             | 18.2497            | 18.3144            | 0.9965  |
| nvidia_deeprecommender         | 11.4864            | 11.5554            | 0.9940  |
| opacus_cifar10                 | 20.7520            | 20.9510            | 0.9905  |
| pyhpc_equation_of_state        | 6.9967             | 6.9972             | 0.9999  |
| pyhpc_isoneutral_mixing        | 27.9351            | 27.8914            | 1.0016  |
| pyhpc_turbulent_kinetic_energy | 22.3412            | 21.3129            | 1.0482  |
| pytorch_CycleGAN_and_pix2pix   | 14.6733            | 14.6800            | 0.9995  |
| pytorch_stargan                | 51.5640            | 51.6061            | 0.9992  |
| pytorch_unet                   | 113.0233           | 113.1834           | 0.9986  |
| resnet152                      | 172.9542           | 172.9537           | 1.0000  |
| resnet18                       | 7.1986             | 7.1692             | 1.0041  |
| resnet50                       | 66.2987            | 66.3271            | 0.9996  |
| resnext50_32x4d                | 62.5773            | 55.4107            | 1.1293  |
| shufflenet_v2_x1_0             | 33.0737            | 32.9855            | 1.0027  |
| soft_actor_critic              | 0.2517             | 0.2538             | 0.9917  |
| speech_transformer             | 21.2798            | 21.2074            | 1.0034  |
| squeezenet1_1                  | 6.3501             | 6.3379             | 1.0019  |
| tacotron2                      | 1095.2057          | 1095.4023          | 0.9998  |
| timm_efficientnet              | 55.2628            | 55.8831            | 0.9889  |
| timm_regnet                    | 212.7725           | 213.2575           | 0.9977  |
| timm_resnest                   | 43.5706            | 43.5622            | 1.0002  |
| timm_vision_transformer        | 10.7047            | 10.7046            | 1.0000  |
| timm_vision_transformer_large  | 378.2258           | 377.8992           | 1.0009  |
| timm_vovnet                    | 81.9470            | 82.6849            | 0.9911  |
| tts_angular                    | 13.3451            | 13.3499            | 0.9996  |
| vgg16                          | 15.5082            | 15.5191            | 0.9993  |
| yolov3                         | 83.8152            | 84.1460            | 0.9961  |

</details>

-----

After looking at the results, I guess it's best to go with the if-chain version. Let me know your thoughts.

(FYI: I will be off until May 1)

ezyang commented:
Alright let's if chain

voznesenskym commented:
> (FYI: I will be off until May 1)

If need be, I can help finish this PR and land it under your attribution.

voznesenskym commented:
> I think a useful thing to know is what the performance difference between:
> 
> ```
> return a and b and c
> ```
> 
> vs
> 
> ```
> if not a:
>   return False
> if not b:
>   return False
> if not c:
>   return False
> return True
> ```
> 
> is. If there is not much difference, then the easiest thing to do is to stop compiling guards as a big list of `and` statements and then you can CSE at the first point a subexpression is used, without requiring to reorder things.

https://github.com/pytorch/pytorch/pull/98374

This PR does that. 

I did some simple local testing, and the perf seems to be identical.

lezcano commented:
>If need be, I can help finish this PR and land it under your attribution

Yukio, as discussed, will be out until the next of the month. If you want this in now now, I think it's fine for you to finish this one up and land it as a PR by yukio and you.

ezyang commented:
SGTM

ysiraichi committed:
Using `ALLOWED_NODE_TYPES` instead of `IGNORED`.

ysiraichi commented:
> https://github.com/pytorch/pytorch/pull/98374
>
> This PR does that.
>
> I did some simple local testing, and the perf seems to be identical.

Apparently that one got obsoleted, and the if-chain went away. Therefore, assuming we are still sticking to if-chains, I'm going to implement them in this PR.

@voznesenskym @ezyang 
Let me know if I am missing anything.

ysiraichi committed:
Refactor CSE classes + if-chains.

ysiraichi committed:
Add test for CSE.

ezyang commented:
@ysiraichi, if you can, grab @anijain2305 to get up to date about what the current state is post your vacation.

ysiraichi commented:
@anijain2305 I sent you a message on Slack. Let me know if you prefer somewhere else.

ysiraichi committed:
Tests fixed and `_ast_unparse_implemented` removed.

ysiraichi committed:
Fix lint problems.

anijain2305 commented on torch/_dynamo/guards.py:
```diff
@@ -632,6 +656,105 @@ def _produce_guard_code(
+            if isinstance(node, PyExprCSEPass.ALLOWED_NODE_TYPES):
+                expr = _ast_unparse(node)
+                self._config.expr_count[expr] = (
+                    self._config.expr_count.setdefault(expr, 0) + 1
```
How about using a default dict?

anijain2305 commented on torch/_dynamo/guards.py:
```diff
@@ -849,6 +972,35 @@ def id_ref(self, obj):
+    codebuf.writeline(f"def ___make_guard_fn({closure_args}):")
+
+    if HAS_UNPARSE_FUNCTIONS:
+        csepass = PyExprCSEPass()
```
Missing `csepass` assignment for the false branch.

ysiraichi replied:
Is it necessary? I only use it under the `HAS_UNPARSE_FUNCTIONS` condition.

ysiraichi commented:
I have started a new perf run: https://github.com/pytorch/pytorch/actions/runs/4874949423

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -814,15 +936,16 @@ def convert(size_or_stride):
+
         if os.environ.get("TORCHDYNAMO_PRINT_GUARDS", None) == "1":
-            print("GUARDS", code)
+            print("GUARDS", " and ".join(unique_code_parts))
```
TBH we should just print the if sled now...

ezyang replied:
Let's do (2); (1) is probably ok too but the outer goop is probably not enlightening. Not (3) because you want to print what actually gets executed.

ezyang reviewed (approved):
go go go

ysiraichi commented:
Here are the results of the dashboard, comparing with https://github.com/pytorch/pytorch/commit/c6c92583576880a61bd6e38a4f2093bac0c1d69c (latest commit up to the base of this PR).

![speedup](https://user-images.githubusercontent.com/3337141/236290184-a9be24f2-df4d-4668-af93-9a4e0695e93a.png)

@ezyang @anijain2305
Compilation time seems to be increased by 1~3 seconds for all 'compiler x benchmark' combinations. Let me know what you think.

anijain2305 commented:
This is the dashboard comparison link - hud.pytorch.org/benchmark/compilers?startTime=Thu%2C%2027%20Apr%202023%2020%3A02%3A31%20GMT&stopTime=Thu%2C%2004%20May%202023%2020%3A02%3A31%20GMT&suite=torchbench&mode=training&dtype=amp&lBranch=ysiraichi/guards-cse&lCommit=36c5a6f70d9e45500d71cb03e5dcd84495c8bae7&rBranch=main&rCommit=c6c92583576880a61bd6e38a4f2093bac0c1d69c

Imo, 1-3 sec increase in compilation time is ok. I am more concerned about the drop in perf for HF models. Almost all HF models see a 1% in perf. TIMM sees a pretty good speedup on the other hand.

Do we know why that is case @ysiraichi? 
@ezyang @voznesenskym Any comments on this?

ysiraichi commented:
Let me look into that.

ysiraichi commented:
I can think of a few things:
- Changed the threshold for CSE-ing an expression to 3 (heard that 2 was not worth it -- not sure who was the source)
- Changed the logic of the nodes that the pass should look into (went from `IGNORED_NODE_TYPES` to `ALLOWED...`)
- I hadn't profiled the if-chain using the perf run (I ran torchbench locally) 

I'm thinking that I should start perf runs for these variants... What do you think?

anijain2305 commented:
> 

Yes, that would be good. We could also take one of the HF models seeing 2% degradation and check locally as well?

ysiraichi commented:
Just to be sure it's not from rebasing, I have started a perf run of the first version of this PR: https://github.com/pytorch/pytorch/actions/runs/4887668049

ezyang commented:
> Changed the threshold for CSE-ing an expression to 3 (heard that 2 was not worth it -- not sure who was the source)

I would disagree with this. 2 should definitely be worth it.

ysiraichi committed:
Generate if-chain separately.

ysiraichi commented:
@anijain2305 @ezyang 
Last week I triggered a perf run on the rebased old version of this PR ([the one that performed better](https://github.com/pytorch/pytorch/pull/98488#issuecomment-1500723149)). In summary, we can now compare the results of the:

- Rebased first version of this PR ([link](hud.pytorch.org/benchmark/compilers?startTime=Mon%2C%2001%20May%202023%2019%3A06%3A16%20GMT&stopTime=Mon%2C%2008%20May%202023%2019%3A06%3A16%20GMT&suite=torchbench&mode=training&dtype=amp&lBranch=ysiraichi/guards-cse&lCommit=a879783db391d076fcc02347e317050dde6a592e&rBranch=main&rCommit=c6c92583576880a61bd6e38a4f2093bac0c1d69c))
- The newest version of this PR ([link](hud.pytorch.org/benchmark/compilers?startTime=Mon%2C%2001%20May%202023%2019%3A06%3A16%20GMT&stopTime=Mon%2C%2008%20May%202023%2019%3A06%3A16%20GMT&suite=torchbench&mode=training&dtype=amp&lBranch=ysiraichi/guards-cse&lCommit=36c5a6f70d9e45500d71cb03e5dcd84495c8bae7&rBranch=main&rCommit=c6c92583576880a61bd6e38a4f2093bac0c1d69c))

As far as I can see, there is not much difference between them. This shows that, probably, CSE effect was reduced after updating this PR.

That said, the slow downs are still a puzzle. I will keep on investigating them.

ysiraichi committed:
Fix tests.

ysiraichi committed:
Reduce use threshold.

ezyang commented:
One possible reason why you are not benefiting as much is if we fixed graph breaks; that will reduce the contribution of guards. I think probably I'd focus on figuring out if the slowdowns are valid or not; if they're within noise we should land it anyway.

ysiraichi commented:
I tried running the inference pass for the huggingface benchmark set locally, but could not reproduce the slow downs. One difference is that I ran this PR (`cse`) against the current base commit of this PR (`8556cf208a3`), instead of using `c6c9258` (the baseline in the dashboard). There seemed to be no difference >1%. Should I check the training pass, too?

Let me know what you think.

**Command:**
```bash
$ python benchmarks/dynamo/huggingface.py --cold-start-latency --device cuda --inference --backend inductor --performance --amp
```

<details>
<summary>HF Performance</summary>

| Name                                    | Latency (cse) | Latency (baseline) | Speedup (cse / baseline) |
|-----------------------------------------|--------------:|-------------------:|-------------------------:|
| AlbertForMaskedLM                       |       97.7787 |            98.0174 |                   0.9976 |
| AlbertForQuestionAnswering              |       97.6842 |            97.5757 |                   1.0011 |
| AllenaiLongformerBase                   |       35.8791 |            35.8827 |                   0.9999 |
| BartForCausalLM                         |       31.4006 |            31.4547 |                   0.9983 |
| BartForConditionalGeneration            |       40.6628 |            40.6341 |                   1.0007 |
| BertForMaskedLM                         |       32.2703 |            32.2552 |                   1.0005 |
| BertForQuestionAnswering                |       25.8519 |            25.7707 |                   1.0032 |
| BlenderbotForCausalLM                   |       22.9920 |            23.0112 |                   0.9992 |
| BlenderbotSmallForCausalLM              |       16.8877 |            16.9482 |                   0.9964 |
| BlenderbotSmallForConditionalGeneration |       32.6727 |            32.5439 |                   1.0040 |
| CamemBert                               |       32.5567 |            32.6716 |                   0.9965 |
| DebertaForMaskedLM                      |       18.2995 |            18.3753 |                   0.9959 |
| DebertaForQuestionAnswering             |       25.4861 |            25.4816 |                   1.0002 |
| DebertaV2ForMaskedLM                    |       22.7355 |            22.8053 |                   0.9969 |
| DebertaV2ForQuestionAnswering           |       18.7498 |            18.7575 |                   0.9996 |
| DistilBertForMaskedLM                   |       32.9714 |            33.0346 |                   0.9981 |
| DistilBertForQuestionAnswering          |       38.8450 |            38.8282 |                   1.0004 |
| DistillGPT2                             |       23.4698 |            23.4863 |                   0.9993 |
| ElectraForCausalLM                      |       19.5771 |            19.5967 |                   0.9990 |
| ElectraForQuestionAnswering             |       25.6321 |            25.6516 |                   0.9992 |
| GPT2ForSequenceClassification           |       17.1804 |            17.1423 |                   1.0022 |
| GoogleFnet                              |       19.3235 |            19.4007 |                   0.9960 |
| LayoutLMForMaskedLM                     |       32.3386 |            32.2846 |                   1.0017 |
| LayoutLMForSequenceClassification       |       25.8167 |            25.8316 |                   0.9994 |
| M2M100ForConditionalGeneration          |       28.7866 |            28.7970 |                   0.9996 |
| MBartForCausalLM                        |       32.0356 |            32.0441 |                   0.9997 |
| MBartForConditionalGeneration           |       41.3213 |            41.3155 |                   1.0001 |
| MT5ForConditionalGeneration             |       12.4830 |            12.4764 |                   1.0005 |
| MegatronBertForCausalLM                 |       24.6044 |            24.5585 |                   1.0019 |
| MegatronBertForQuestionAnswering        |       42.0026 |            41.9453 |                   1.0014 |
| MobileBertForMaskedLM                   |       41.0459 |            41.1578 |                   0.9973 |
| MobileBertForQuestionAnswering          |       27.7888 |            27.6717 |                   1.0042 |
| OPTForCausalLM                          |       27.8872 |            27.8946 |                   0.9997 |
| PLBartForCausalLM                       |       26.8276 |            26.8591 |                   0.9988 |
| PLBartForConditionalGeneration          |       28.5260 |            28.4547 |                   1.0025 |
| PegasusForCausalLM                      |       23.4089 |            23.4320 |                   0.9990 |
| PegasusForConditionalGeneration         |       49.5742 |            49.5978 |                   0.9995 |
| RobertaForCausalLM                      |       32.2950 |            32.3064 |                   0.9996 |
| RobertaForQuestionAnswering             |       25.7564 |            25.7716 |                   0.9994 |
| Speech2Text2ForCausalLM                 |       15.2806 |            15.2656 |                   1.0010 |
| T5ForConditionalGeneration              |       18.4347 |            18.4365 |                   0.9999 |
| T5Small                                 |       18.4758 |            18.4288 |                   1.0026 |
| TrOCRForCausalLM                        |       46.7501 |            46.7689 |                   0.9996 |
| XGLMForCausalLM                         |       15.8072 |            15.8674 |                   0.9962 |
| XLNetLMHeadModel                        |       64.7978 |            64.4825 |                   1.0049 |
| YituTechConvBert                        |       36.4039 |            36.3920 |                   1.0003 |

</details>

ezyang commented:
OK, maybe you got unlucky on your run and someone committed a regression. I recommend rebasing this PR onto a commit which we have a perf run for (at time of writing, that's 4c3b52a5a99e8aed2576773829e6a353a65aa2b4) and then redoing the experiment on GCP.

Or we can just land it. You have my approve.

ysiraichi commented:
Running: https://github.com/pytorch/pytorch/actions/runs/4939065921

ysiraichi commented:
I think the last perf run was affected by the GitHub outage. 
Many of the HF benchmarks [were marked as `passed -> undefined`](https://hud.pytorch.org/benchmark/huggingface/inductor_with_cudagraphs?startTime=Thu,%2004%20May%202023%2017:22:52%20GMT&stopTime=Thu,%2011%20May%202023%2017:22:52%20GMT&mode=training&dtype=amp&lBranch=ysiraichi/guards-cse&lCommit=d441c114e91915e5f9d28dcfddb47177fb6a1335&rBranch=main&rCommit=4c3b52a5a99e8aed2576773829e6a353a65aa2b4). However, looking at [the raw log files](https://ossci-raw-job-status.s3.amazonaws.com/log/13380133047), there seemed to be no problem.
So, I fired up another perf run for the same commit: https://github.com/pytorch/pytorch/actions/runs/4950341860

ysiraichi committed:
Fix lint.

desertfire commented:
I cancelled your latest dashboard run due to some dashboard upgrading. I will resubmit for you after that's done.

desertfire commented:
> I cancelled your latest dashboard run due to some dashboard upgrading. I will resubmit for you after that's done.

OK, dashboard is ready but you need to rebase your `ysiraichi/guards-cse` branch in order to submit a perf test. I don't want do that for you in case it messes up with anything you are working on. Please do it at your convenience.

ysiraichi commented:
Running: https://github.com/pytorch/pytorch/actions/runs/4980506650

ysiraichi commented:
[Results are out!](https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2009%20May%202023%2018%3A23%3A10%20GMT&stopTime=Tue%2C%2016%20May%202023%2018%3A23%3A10%20GMT&suite=torchbench&mode=training&dtype=amp&lBranch=ysiraichi/guards-cse&lCommit=55b9e6cdbbed7c38edae7847a38bd1c151c7ed77&rBranch=main&rCommit=7dd8e08817ee59c926922409062e25f30408469b)

![speedup](https://github.com/pytorch/pytorch/assets/3337141/c8484cd5-5b8a-4260-ba75-9d478d637a2e)

@ezyang @anijain2305 
Is it ok if I merge this once this PR turns green?

lezcano commented:
@pytorchbot merge

ysiraichi commented:
@pytorchbot merge

malfet commented:
This regressed MacOS x86 builds, and in general, string comparison feels like a very fragile thing to do.
I.e. on MacOS it for some reason `test_guard_function_builder_with_cse` generates:
```
def ___make_guard_fn():
    def guard(L):
        if not (x[0].a < x[1].a * (3 - x[2].a)):
            return False
        if not (a.b.c[0].d.e + a.b.c[1].d.e * a.b.c[2].d.e > 0):
            return False
        if not (f(m.n[0], '0').x.y.z * f(m.n[0], '1').x.y.z * f(m.n[0], '2').x.y.z < 512):
            return False
        if not (self.g(a, b).k + (1 - self.g(a, b).k) <= m[0].a + self.g(a, b).k):
            return False
        return True
    return guard
```
which is indential to the one one is trying to match, but without any intermediate wars

ysiraichi commented:
Ah, right. Good catch. Thanks, @malfet .

malfet commented:
Hmm, looks like the way code is written right now, it would not really work without astunparse, should we just add it as mandatory dependency on python-3.8?

ysiraichi commented:
How so? To me, it looks like it would work without `astunparse`.

malfet commented:
@ysiraichi do you see where this call is guarded by `HAS_UNPARSE_FUNCTIONS == True`
https://github.com/pytorch/pytorch/blob/556bb691fd0887f9756625e5727336fdb8c30914/torch/_dynamo/guards.py#L735

ysiraichi commented:
The object instantiation is guarded.
https://github.com/pytorch/pytorch/blob/556bb691fd0887f9756625e5727336fdb8c30914/torch/_dynamo/guards.py#L1024

malfet commented:
Ok, so I'll just guard this test than...

bobrenjc93 commented:
@ysiraichi this can run into max recursion depth issues for sufficiently complex sympy guards. any chance you can take a look? here's a repro: 

```
    def test_finite_recursion(self):
        from torch._dynamo.guards import PyExprCSEPass

        testcase = self.CSETestCase
        testcases = [
            # Nothing gets CSE-d, since the only repeated sub-expression is 'x'.
            # i.e. not a node type we are interested on.
            testcase(expr="s0 + s1 + 12*s10 + 6*s100 + s1000 + s1001 + s1002 + s1003 + s1004 + s1005 + s1006 + s1007 + s1008 + s1009 + s101 + s1010 + s1011 + s1012 + s1013 + s1014 + s1015 + s1016 + s1017 + s1018 + s1019 + s102 + s1020 + s1021 + s1022 + s1023 + s1024 + s1025 + s1026 + s1027 + s1028 + s1029 + s103 + s1030 + s1031 + s1032 + s1033 + s1034 + s1035 + s1036 + s1037 + s1038 + s1039 + s104 + s1040 + s1041 + s1042 + s1043 + s1044 + s1045 + s1046 + s1047 + s1048 + s1049 + s105 + s1050 + s1051 + s1052 + s1053 + s1054 + s1055 + s1056 + s1057 + s1058 + s1059 + s106 + s1060 + s1061 + s1062 + s1063 + s1064 + s1065 + s1066 + s1067 + s1068 + s1069 + s107 + s1070 + s1071 + s1072 + s1073 + s1074 + 3*s1075 + 3*s1076 + s1077 + s1078 + s1079 + s108 + s1080 + s1081 + s1082 + s1083 + s1084 + s1085 + s1086 + s1087 + s1088 + s1089 + s109 + s1090 + s1091 + s1092 + s1093 + s1094 + s1095 + s1096 + s1097 + s1098 + s1099 + s11 + s110 + s1100 + s1101 + s1102 + s1103 + s1104 + s1105 + s1106 + s1107 + s1108 + s1109 + s111 + s1110 + s1111 + s1112 + s1113 + s1114 + s1115 + s1116 + s1117 + s1118 + s1119 + s112 + s1120 + s1121 + s1122 + s1123 + s1124 + s1125 + s1126 + s1127 + s1128 + s1129 + s113 + s1130 + s1131 + s1132 + s1133 + s1134 + s1135 + s1136 + s1137 + s1138 + s1139 + 12*s114 + s1140 + s1141 + s1142 + s1143 + s1144 + s1145 + s1146 + s1147 + s1148 + s1149 + 12*s115 + s1150 + s1151 + s1152 + s1153 + s1154 + s1155 + s1156 + s1157 + s1158 + s1159 + 8*s116 + s1160 + s1161 + s1162 + s1163 + s1164 + s1165 + s1166 + s1167 + 3*s1168 + 3*s1169 + 8*s117 + 3*s1170 + s1171 + s1172 + s1173 + s1174 + s1175 + s1176 + s1177 + s1178 + s1179 + 6*s118 + s1180 + s1181 + s1182 + s1183 + s1184 + s1185 + s1186 + s1187 + s1188 + s1189 + s119 + s1190 + s1191 + s1192 + s1193 + s1194 + s1195 + s1196 + s1197 + s1198 + s1199 + s12 + s120 + s1200 + s1201 + s1202 + s1203 + s1204 + s1205 + s1206 + s1207 + s1208 + s1209 + s121 + s1210 + s1211 + s1212 + s1213 + s1214 + s1215 + s1216 + s1217 + s1218 + s1219 + s122 + s1220 + s1221 + s1222 + s1223 + s1224 + s1225 + s1226 + s1227 + s1228 + s1229 + s123 + s1230 + s1231 + s1232 + s1233 + s1234 + s1235 + s1236 + s1237 + s1238 + s1239 + s124 + s1240 + s1241 + s1242 + s1243 + s1244 + s1245 + s1246 + s1247 + s1248 + s1249 + s125 + s1250 + s1251 + s1252 + s1253 + s1254 + s1255 + s1256 + s1257 + s1258 + s1259 + s126 + s1260 + s1261 + 3*s1262 + 3*s1263 + 3*s1264 + s1265 + s1266 + s1267 + s1268 + s1269 + s127 + s1270 + s1271 + s1272 + s1273 + s1274 + s1275 + s1276 + s1277 + s1278 + s1279 + s128 + s1280 + s1281 + s1282 + s1283 + s1284 + s1285 + s1286 + s1287 + s1288 + s1289 + s129 + s1290 + s1291 + s1292 + s1293 + s1294 + s1295 + s1296 + s1297 + s1298 + s1299 + s13 + s130 + s1300 + s1301 + s1302 + s1303 + s1304 + s1305 + s1306 + s1307 + s1308 + s1309 + 6*s131 + s1310 + s1311 + s1312 + s1313 + s1314 + s1315 + s1316 + s1317 + s1318 + s1319 + 6*s132 + s1320 + s1321 + s1322 + s1323 + s1324 + s1325 + s1326 + s1327 + s1328 + s1329 + 6*s133 + s1330 + s1331 + s1332 + s1333 + s1334 + s1335 + s1336 + s1337 + s1338 + s1339 + 6*s134 + s1340 + s1341 + s1342 + s1343 + s1344 + s1345 + s1346 + s1347 + s1348 + s1349 + 3*s135 + s1350 + s1351 + s1352 + s1353 + s1354 + s1355 + s1356 + s1357 + s1358 + s1359 + 3*s136 + s1360 + s1361 + s1362 + s1363 + s1364 + s1365 + s1366 + s1367 + s1368 + s1369 + 3*s137 + s1370 + s1371 + s1372 + s1373 + s1374 + s1375 + s1376 + s1377 + s1378 + s1379 + 3*s138 + s1380 + s1381 + s1382 + s1383 + s1384 + s1385 + s1386 + s1387 + s1388 + s1389 + 3*s139 + s1390 + s1391 + s1392 + s1393 + s1394 + s1395 + s1396 + s1397 + s1398 + s1399 + s14 + 3*s140 + s1400 + s1401 + s1402 + s1403 + s1404 + s1405 + s1406 + s1407 + s1408 + s1409 + 3*s141 + s1410 + s1411 + s1412 + s1413 + s1414 + s1415 + s1416 + s1417 + s1418 + s1419 + 3*s142 + s1420 + s1421 + s1422 + s1423 + s1424 + s1425 + s1426 + s1427 + s1428 + s1429 + 3*s143 + s1430 + s1431 + s1432 + s1433 + s1434 + s1435 + s1436 + s1437 + s1438 + s1439 + 3*s144 + s1440 + s1441 + s1442 + s1443 + s1444 + s1445 + s1446 + s1447 + s1448 + s1449 + 3*s145 + s1450 + s1451 + s1452 + s1453 + s1454 + s1455 + s1456 + s1457 + s1458 + s1459 + 3*s146 + s1460 + s1461 + s1462 + s1463 + s1464 + s1465 + s1466 + s1467 + s1468 + s1469 + s147 + s1470 + s1471 + s1472 + s1473 + s1474 + s1475 + s1476 + s1477 + s1478 + s1479 + s148 + s1480 + s1481 + s1482 + s1483 + s1484 + s1485 + s1486 + s1487 + s1488 + s1489 + s149 + s1490 + s1491 + s1492 + s1493 + s1494 + s1495 + s1496 + s1497 + s1498 + s1499 + s15 + s150 + s1500 + s1501 + s1502 + s1503 + s1504 + s1505 + s1506 + s1507 + s1508 + s1509 + s151 + s1510 + s1511 + s1512 + s1513 + s1514 + s1515 + s1516 + s1517 + s1518 + s1519 + s152 + s1520 + s1521 + s1522 + s1523 + s1524 + s1525 + s1526 + s1527 + s1528 + s1529 + s153 + s1530 + s1531 + s1532 + s1533 + s1534 + s154 + s155 + s156 + s157 + s158 + s159 + s16 + s160 + s161 + s162 + s163 + s164 + s165 + s166 + s167 + s168 + s169 + s17 + s170 + s171 + s172 + s173 + s174 + s175 + s176 + s177 + s178 + s179 + 6*s18 + s180 + s181 + s182 + s183 + s184 + s185 + s186 + s187 + s188 + s189 + s19 + s190 + s191 + s192 + s193 + s194 + s195 + s196 + s197 + s198 + s199 + s2 + 3*s20 + s200 + s201 + s202 + s203 + s204 + s205 + s206 + s207 + s208 + s209 + s21 + s210 + s211 + s212 + s213 + s214 + s215 + s216 + 6*s217 + 6*s218 + 6*s219 + s22 + 3*s220 + 3*s221 + s222 + s223 + s224 + s225 + s226 + s227 + s228 + s229 + 3*s23 + s230 + s231 + s232 + 12*s233 + 12*s234 + 8*s235 + 6*s236 + 3*s237 + 3*s238 + 3*s239 + s24 + 3*s240 + 3*s241 + s242 + s243 + s244 + s245 + s246 + s247 + s248 + s249 + s25 + s250 + s251 + s252 + s253 + s254 + s255 + s256 + s257 + s258 + s259 + 12*s26 + s260 + s261 + s262 + s263 + s264 + s265 + s266 + s267 + s268 + s269 + s27 + s270 + s271 + s272 + s273 + s274 + s275 + s276 + s277 + s278 + s279 + s28 + s280 + s281 + s282 + s283 + s284 + s285 + s286 + s287 + s288 + s289 + 8*s29 + s290 + s291 + s292 + s293 + s294 + s295 + s296 + s297 + s298 + s299 + s3 + s30 + 6*s300 + 3*s301 + 3*s302 + s303 + s304 + s305 + s306 + s307 + s308 + s309 + 3*s31 + s310 + s311 + s312 + s313 + s314 + s315 + s316 + s317 + s318 + s319 + s32 + s320 + s321 + 8*s322 + 3*s323 + 3*s324 + s325 + s326 + s327 + s328 + s329 + s33 + s330 + s331 + s332 + s333 + s334 + 6*s335 + 3*s336 + 3*s337 + 3*s338 + s339 + 3*s34 + s340 + s341 + s342 + s343 + s344 + s345 + s346 + s347 + s348 + s349 + s35 + s350 + s351 + s352 + s353 + s354 + s355 + s356 + s357 + s358 + s359 + 3*s36 + s360 + s361 + s362 + s363 + s364 + s365 + s366 + s367 + s368 + s369 + s37 + s370 + s371 + s372 + s373 + s374 + s375 + s376 + s377 + s378 + s379 + 8*s38 + s380 + s381 + s382 + s383 + s384 + s385 + s386 + s387 + s388 + s389 + 8*s39 + s390 + s391 + s392 + s393 + s394 + s395 + s396 + s397 + 6*s398 + 6*s399 + s4 + s40 + 6*s400 + s401 + s402 + s403 + s404 + s405 + s406 + s407 + s408 + s409 + s41 + s410 + s411 + s412 + s413 + s414 + s415 + s416 + 6*s417 + 6*s418 + s419 + s42 + s420 + s421 + s422 + s423 + s424 + s425 + s426 + s427 + s428 + s429 + s43 + 6*s430 + 3*s431 + 3*s432 + 3*s433 + 3*s434 + 3*s435 + 3*s436 + 3*s437 + 3*s438 + 3*s439 + s44 + 3*s440 + 3*s441 + s442 + s443 + s444 + s445 + s446 + s447 + s448 + s449 + 12*s45 + s450 + s451 + s452 + s453 + s454 + s455 + s456 + s457 + s458 + s459 + s46 + s460 + s461 + s462 + s463 + s464 + s465 + s466 + s467 + s468 + s469 + s47 + s470 + s471 + s472 + s473 + s474 + s475 + s476 + s477 + s478 + s479 + 6*s48 + s480 + s481 + s482 + s483 + s484 + s485 + s486 + s487 + s488 + s489 + 3*s49 + s490 + s491 + s492 + s493 + s494 + s495 + s496 + s497 + s498 + 6*s499 + 8*s5 + s50 + 3*s500 + s501 + s502 + s503 + s504 + s505 + s506 + s507 + s508 + s509 + s51 + s510 + s511 + s512 + s513 + s514 + s515 + 3*s516 + s517 + s518 + s519 + 8*s52 + s520 + s521 + s522 + s523 + s524 + s525 + s526 + s527 + s528 + s529 + 3*s53 + 6*s530 + 6*s531 + 3*s532 + 3*s533 + 3*s534 + 3*s535 + 3*s536 + 3*s537 + 3*s538 + 3*s539 + s54 + 3*s540 + 3*s541 + s542 + s543 + s544 + s545 + s546 + s547 + s548 + s549 + s55 + s550 + s551 + s552 + s553 + s554 + s555 + s556 + s557 + s558 + s559 + 3*s56 + s560 + s561 + s562 + s563 + s564 + s565 + s566 + s567 + s568 + s569 + s57 + s570 + s571 + s572 + s573 + s574 + s575 + s576 + s577 + s578 + s579 + s58 + s580 + s581 + s582 + s583 + s584 + s585 + s586 + s587 + s588 + s589 + s59 + s590 + s591 + s592 + s593 + 3*s594 + 3*s595 + s596 + s597 + s598 + s599 + s6 + s60 + s600 + s601 + s602 + s603 + s604 + s605 + s606 + s607 + s608 + s609 + s61 + s610 + s611 + s612 + s613 + s614 + 3*s615 + s616 + s617 + s618 + s619 + 3*s62 + s620 + s621 + s622 + s623 + s624 + s625 + s626 + s627 + 3*s628 + 3*s629 + s63 + 3*s630 + 3*s631 + s632 + s633 + s634 + s635 + s636 + s637 + s638 + s639 + s64 + s640 + s641 + s642 + s643 + s644 + s645 + s646 + s647 + s648 + s649 + s65 + s650 + s651 + s652 + s653 + s654 + s655 + s656 + s657 + s658 + s659 + s66 + s660 + s661 + s662 + s663 + s664 + s665 + s666 + s667 + s668 + s669 + s67 + s670 + s671 + s672 + s673 + s674 + s675 + s676 + s677 + s678 + s679 + s68 + s680 + s681 + s682 + s683 + s684 + s685 + s686 + s687 + s688 + 3*s689 + 6*s69 + s690 + s691 + s692 + s693 + s694 + s695 + s696 + s697 + s698 + s699 + s7 + 3*s70 + s700 + s701 + s702 + s703 + 3*s704 + 3*s705 + s706 + s707 + s708 + s709 + 3*s71 + s710 + s711 + s712 + s713 + s714 + s715 + s716 + s717 + 3*s718 + 3*s719 + 6*s72 + s720 + s721 + s722 + s723 + s724 + s725 + s726 + s727 + s728 + s729 + 6*s73 + s730 + s731 + s732 + s733 + s734 + s735 + s736 + s737 + s738 + s739 + 8*s74 + s740 + s741 + s742 + s743 + s744 + s745 + s746 + s747 + s748 + s749 + s75 + s750 + s751 + s752 + s753 + s754 + s755 + s756 + s757 + s758 + s759 + s76 + s760 + s761 + s762 + s763 + s764 + s765 + s766 + s767 + s768 + s769 + s77 + s770 + s771 + s772 + s773 + s774 + s775 + s776 + s777 + s778 + s779 + 3*s78 + s780 + s781 + s782 + s783 + s784 + s785 + s786 + s787 + s788 + s789 + s79 + s790 + s791 + s792 + s793 + s794 + s795 + s796 + s797 + s798 + s799 + s8 + s80 + s800 + s801 + s802 + s803 + s804 + s805 + s806 + s807 + s808 + s809 + s81 + s810 + s811 + s812 + s813 + s814 + s815 + s816 + s817 + s818 + s819 + s82 + s820 + s821 + s822 + s823 + s824 + s825 + s826 + s827 + s828 + s829 + 8*s83 + s830 + s831 + s832 + s833 + s834 + s835 + s836 + s837 + s838 + s839 + s84 + s840 + s841 + s842 + s843 + s844 + s845 + s846 + s847 + s848 + s849 + s85 + s850 + s851 + s852 + s853 + s854 + s855 + s856 + s857 + s858 + s859 + s86 + s860 + s861 + s862 + s863 + s864 + s865 + s866 + s867 + s868 + s869 + 3*s87 + s870 + s871 + s872 + s873 + s874 + s875 + s876 + s877 + s878 + s879 + s88 + s880 + s881 + s882 + s883 + s884 + s885 + s886 + s887 + s888 + s889 + s89 + s890 + s891 + s892 + s893 + s894 + s895 + 3*s896 + 3*s897 + 3*s898 + 3*s899 + s9 + s90 + 3*s900 + s901 + s902 + s903 + s904 + s905 + s906 + s907 + s908 + s909 + 3*s91 + s910 + s911 + s912 + s913 + s914 + s915 + s916 + s917 + s918 + s919 + 3*s92 + s920 + s921 + s922 + s923 + s924 + s925 + s926 + s927 + s928 + s929 + 3*s93 + s930 + s931 + s932 + s933 + s934 + s935 + s936 + s937 + s938 + s939 + s94 + s940 + s941 + s942 + s943 + s944 + s945 + s946 + s947 + s948 + s949 + s95 + s950 + s951 + s952 + s953 + s954 + s955 + s956 + s957 + s958 + s959 + s96 + s960 + 3*s961 + s962 + s963 + s964 + s965 + s966 + s967 + s968 + s969 + s97 + s970 + s971 + s972 + s973 + s974 + s975 + 3*s976 + s977 + s978 + s979 + s98 + s980 + s981 + s982 + s983 + s984 + s985 + s986 + s987 + s988 + s989 + 3*s99 + s990 + s991 + s992 + s993 + s994 + s995 + s996 + s997 + s998 + s999 == 0"),
        ]
        csepass = PyExprCSEPass()
        csepass.count([t.expr for t in testcases])

        for t in testcases:
            preface, expr = csepass.replace(t.expr)
            self.assertEqual(preface, t.preface)
            expected = t.expected if t.expected is not None else t.expr
            self.assertEqual(expr, expected)
```
